### ISSUE TITLE
reana: create example workflow and run reco on 2011 sample

### ISSUE THREAD

**Issue by tiborsimko** [reanahub/reana-demo-cms-reco#8]
Following the test in #5  for some CMS 2011 RAW sample, create `reana.yaml` and run the process through REANA using CVMFS.  Compare the obtained AOD resulting from VM run and from REANA run.

**Comment by dprelipcean** [comment]
Steps taken:

- create [configuration file](https://github.com/dprelipcean/reana-demo-cms-reco/blob/reana_example_2011/reco_cmsdriver2011.py) as specified by the [CMS legacy data validation](https://github.com/cms-legacydata-validation/RAWToAODValidation/tree/2011)
- ToDo: run the reconstruction, both on VM and reana
- ToDo: run the analyzer between the two, and compared to Open Data AOD file

**Comment by dprelipcean** [comment]
@katilp When running this [demo analyzer](https://github.com/cms-opendata-validation/2011-minimumbias-compare-rereco/blob/master/demoanalyzer_cfg.py) (adapted to my files) to validate the outputs, I get the following error, both on my CMS VM and in reana (I created a separate [workflow](https://github.com/dprelipcean/reana-demo-cms-reco/blob/reana_example_2011/reana-analyzer.yaml) to isolate the issue):

```console
----- Begin Fatal Exception 19-Aug-2019 14:26:09 CEST-----------------------
An exception of category 'PluginNotFound' occurred while
   [0] Constructing the EventProcessor
Exception Message:
Unable to find plugin 'PhysicsObjectsHistos'. Please check spelling of name.
```
I tried to put the `PhysicsObjectsHistos.cc` file in the same directory, still with no success.

Is there something I am missing? I am following the guidelines from [this repo](https://github.com/cms-legacydata-validation/RAWToAODValidation/tree/2011).

**Comment by clelange** [comment]
@dprelipcean Not sure if you solved this already, but here's an explanation: CMSSW code needs to live in a certain directory structure w.r.t. the CMSSW release set up. For example, when setting up `CMSSW_5_3_32` in `/code`, `PhysicsObjectsHistos.cc` needs to be in `/code/CMSSW_5_3_32/src/RAWToAODValidation/ElectronValidation/src/PhysicsObjectsHistos.cc`, where `RAWToAODValidation` and `ElectronValidation` are just arbitrary directory names, but there need to be two such subdirectories for the code to be compiled by `scram b` in the first place. Otherwise, you will see that `scram` basically does nothing. Hope that helps.

**Comment by dprelipcean** [comment]
@clelange Thanks! interesting project structure; I am now able to run the [demoanalyzer.py](https://github.com/dprelipcean/reana-demo-cms-reco/blob/reana_example_2011/demoanalyzer_cfg.py) inside the CMS VM, but not in reana using [this workflow](https://github.com/dprelipcean/reana-demo-cms-reco/blob/reana_example_2011/reana-analyzer.yaml)

The error I get is:

```console
 fatal error: boost/shared_ptr.hpp: No such file or directory
```
Do you think it's possbile that the `cmsopendata/cmssw_5_3_32` docker image does not contain all dependencies/headers?

Full stack trace:
```console
WARNING: In non-interactive mode release checks e.g. deprecated releases, production architectures are disabled.
>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Entering Package CMSreco/validation
>> Creating project symlinks
Entering library rule at CMSreco/validation
>> Compiling  /var/reana/users/00000000-0000-0000-0000-000000000000/workflows/efd42bab-90c3-4308-ae84-73fb31e63960/CMSSW_5_3_32/src/CMSreco/validation/src/PhysicsObjectsHistos.cc 
In file included from /opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/FWCore/Framework/interface/Frameworkfwd.h:10:0,
                 from /var/reana/users/00000000-0000-0000-0000-000000000000/workflows/efd42bab-90c3-4308-ae84-73fb31e63960/CMSSW_5_3_32/src/CMSreco/validation/src/PhysicsObjectsHistos.cc:18:
/opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/DataFormats/Common/interface/EDProductfwd.h:9:32: fatal error: boost/shared_ptr.hpp: No such file or directory
compilation terminated.
In file included from /opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/FWCore/Framework/interface/Frameworkfwd.h:10:0,
                 from /var/reana/users/00000000-0000-0000-0000-000000000000/workflows/efd42bab-90c3-4308-ae84-73fb31e63960/CMSSW_5_3_32/src/CMSreco/validation/src/PhysicsObjectsHistos.cc:18:
/opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/DataFormats/Common/interface/EDProductfwd.h:9:32: fatal error: boost/shared_ptr.hpp: No such file or directory
compilation terminated.
gmake: *** [tmp/slc6_amd64_gcc472/src/CMSreco/validation/src/CMSrecovalidation/PhysicsObjectsHistos.o] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

**Comment by clelange** [comment]
@dprelipcean It is possible, but in principle everything should be there. Could you just recursively search `/opt/cms` for the missing files? I think it could just be an issue with the `BuildFile.xml` of the validation package.
Mind also that convention in CMS is to use CamelCase (it won't affect the compilation, but it's better to be consistent), i.e. I would suggest you use `Reconstruction/Validation` as directory structure (don't need "CMS" in the name since we know that it is ;-) ).

**Comment by dprelipcean** [comment]
@clelange 
```console
cmsusr@5234750db1c4 /opt/cms $  /opt/cms $ find . -name "shared_ptr.hpp"
./slc6_amd64_gcc472/external/boost/1.47.0-cms/include/boost/shared_ptr.hpp
./slc6_amd64_gcc472/external/boost/1.47.0-cms/include/boost/smart_ptr/shared_ptr.hpp
./slc6_amd64_gcc472/external/boost/1.47.0-cms/include/boost/serialization/shared_ptr.hpp
./slc6_amd64_gcc472/external/boost/1.47.0-cms/include/boost/asio/detail/shared_ptr.hpp
./slc6_amd64_gcc472/external/boost/1.47.0-cms/include/boost/interprocess/smart_ptr/shared_ptr.hpp
```
and the `BuildFile.xml` looks like:
```console
cmsusr@5234750db1c4 /opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/DataFormats/Common $ cat BuildFile.xml 
<use   name="DataFormats/Provenance"/>
<use   name="FWCore/MessageLogger"/>
<use   name="FWCore/Utilities"/>
<use   name="DataFormats/StdDictionaries"/>
<use   name="boost"/>
<use   name="rootrflx"/>
<export>
  <lib   name="1"/>
</export>
```

I assume this searches for `boost` inside `/opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/` and not in `/opt/cms/slc6_amd64_gcc472/external/boost/1.47.0-cms/include/`, hence the failure?

**Comment by clelange** [comment]
The `boost` libraries should also be in the environment. I think the easiest way to understand this would be to just print `env | grep boost` (or without `grep boost` to be safe to catch everything) for the VM and the container when running in REANA, and figure out the differences (apart from the different base paths `/opt/cms` vs. `/cvmfs/cms.cern.ch`...).

**Comment by dprelipcean** [comment]
Hei @clelange, thanks a lot for all the quick feedback!
I had a look at the `boost` libraries both in the VM and in the image, and there was no difference (apart from base paths).

Fooling around a bit and looking at the [cms validation](https://github.com/cms-legacydata-validation/RAWToAODValidation/tree/2011), I noticed that one actually needs not only 2, but 3 level subdirectories in the image, i.e.:
```console
~/CMSSW_5_3_32/src/WorkDir/Reconstruction/Validation $ scram b
>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Nothing to be done for src/WorkDir/Reconstruction/Validation
>> Local Products Rules ..... started
>> Local Products Rules ..... done
gmake[1]: Entering directory `/home/cmsusr/CMSSW_5_3_32'
>> Creating project symlinks
>> Done python_symlink
>> Compiling python modules python
>> All python modules compiled
>> Pluging of all type refreshed.
>> Done generating edm plugin poisoned information
gmake[1]: Leaving directory `/home/cmsusr/CMSSW_5_3_32'
```
whereas:
```console
~/CMSSW_5_3_32/src/WorkDir/Reconstruction $ scram b
Reading cached build data
>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Entering Package WorkDir/Reconstruction
>> Creating project symlinks
Entering library rule at WorkDir/Reconstruction
>> Compiling  /home/cmsusr/CMSSW_5_3_32/src/WorkDir/Reconstruction/src/PhysicsObjectsHistos.cc 
In file included from /opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/FWCore/Framework/interface/Frameworkfwd.h:10:0,
                 from /home/cmsusr/CMSSW_5_3_32/src/WorkDir/reana-demo-cms-reco/src/PhysicsObjectsHistos.cc:18:
/opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/DataFormats/Common/interface/EDProductfwd.h:9:32: fatal error: boost/shared_ptr.hpp: No such file or directory
compilation terminated.
In file included from /opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/FWCore/Framework/interface/Frameworkfwd.h:10:0,
                 from /home/cmsusr/CMSSW_5_3_32/src/WorkDir/Reconstruction/src/PhysicsObjectsHistos.cc:18:
/opt/cms/slc6_amd64_gcc472/cms/cmssw/CMSSW_5_3_32/src/DataFormats/Common/interface/EDProductfwd.h:9:32: fatal error: boost/shared_ptr.hpp: No such file or directory
compilation terminated.
gmake: *** [tmp/slc6_amd64_gcc472/src/WorkDir/Reconstruction/src/WorkDirReconstruction/PhysicsObjectsHistos.o] Error 1
```
Previously, in the VM, I needed only 2.

Or is this
```console
>> Nothing to be done for src/WorkDir/Reconstruction/Validation
```
 what you meant by 

> Otherwise, you will see that scram basically does nothing.

**Comment by dprelipcean** [comment]
@clelange I realized that my mistake was that I didn't have a `BuildFile.xml` to start with, as I didn't know it is needed for the validation. I will now see what I have to put in it to make it work.
Thanks for the help and sorry for the newbee mistake!

**Comment by dprelipcean** [comment]
Closed by #18
